src/ctia/bundle/core.clj

```diff
@@ -340,155 +336,165 @@
+             That means the last partition in result list is incomplete and we must use `offset` to fetch next page."
+  ([f store filter-map identity-map params-map]
+   (scroll-with-limit [] [] f store filter-map identity-map params-map))
+  ([acc prev-data f store filter-map identity-map {:keys [limit] :as params-map}]
+   (let [{data :data {{:keys [offset search_after] :as next} :next} :paging}
+         (store/list-records store filter-map identity-map params-map)
+         params-map (select-keys params-map [:limit :sort])]
+     (if (nil? next)
+       (into acc (take limit (concat prev-data data)))
+       (let [[chunk & chunks] (partition-by f data)]
+         (if (empty? chunks)
+           (recur (into acc (take limit (concat prev-data chunk)))
+                  [] f store filter-map identity-map
+                  (assoc params-map :search_after search_after))
+           (recur (reduce into acc (cons (take limit (concat prev-data chunk)) (butlast chunks)))
+                  (last chunks) f store filter-map identity-map
+                  (assoc params-map :offset offset))))))))
+
+(defn relationships-filter [records related-to source-type target-type]
+  {:query (cond->> [(format "%s:(%s)" (name related-to) (string/join " OR " (map #(format "\"%s\"" %) (map :id records))))]
+            source-type (cons (format "source_ref:*%s*" (name source-type)))
+            target-type (cons (format "target_ref:*%s*" (name target-type)))
+            :always (reverse)
+            :always (string/join " AND "))})
+
+(defn fetch-relationships [records identity-map
+                           {:keys [related_to
+                                   source_type
+                                   target_type]
+                            :or {related_to [:source_ref :target_ref]}}
+                           {{:keys [get-store]} :StoreService
+                            {:keys [get-in-config]} :ConfigService
+                            {:keys [send-event]} :RiemannService}]
+  (when (seq records)
     (let [start (System/currentTimeMillis)
-          res (-> (get-store (keyword entity-type))
-                  (read-fn
-                   id
-                   identity-map
-                   {}))]
-      (send-event {:service "Export bundle fetch record"
+          store (get-store :relationship)
+          limit (get-in-config [:ctia :http :bundle :export :max-relationships] 1000)
+          res (into #{}
+                    (mapcat #(scroll-with-limit % store
+                                                (relationships-filter records % source_type target_type)
+                                                identity-map
+                                                {:limit limit
+                                                 :sort [{"timestamp" "desc"}
+                                                        {(name %) "desc"}]}))
```

Comment:
Nice one. I've missed that corner case. thanks

Reply:
Tests reveal id is not needed in `:sort`.

I'm currently working on a more representative test case that should also cover this corner case. But for now I will remove id from sorting set.

Reply:
I've tried adding "_shard_doc" into the set but I've got `"reason": "No mapping found for [_shard_doc] in order to sort on"`

Reply:
Sorry, I didn't realize that it already contains `source_ref` or `target_ref` as tiebreaker

Reply:
The doc is for ES7, `_shard_doc` does not exists in ES5.
Recommendation for ES5, using _doc does not work indeed (probably the same thing) because _doc is unique per shard.
In tests it's hard to reproduce: add relationships with same source_ref or target_ref with a fixed timestamp and limit 1, you should hit the problem.

Reply:
tiebreaker is important when `search_after` is used as a pagination mechanism to retrieve all the results. for `scroll-with-limit`, it is acceptable to not use it at all. The algorithm uses `search_after` only when relationships are fully collected for some `source_ref` of `target_ref` and we must jump to some page of search results that begins with another ref. There is a test case that covers this behaviour.

Reply:
Ok great approach to overcome the limitation of `size + offset`  of ES that works when we limit the number of sources.